More economical fee estimates for RBF and RPC options to control

morcos opened the pull request:
This PR takes advantage of the new fee estimation feature that will potentially give lower estimates if recent market conditions warrant it. The logic used here is that any time a transaction signals opt-in-RBF and uses automatic fee estimation then it will use the non-conservative estimate. Transactions which do not signal opt-in-RBF will still use the default conservative estimate.

In a nutshell conservative estimates require that your fee rate would meet the necessary confirmation threshold for double your target at longer time horizons as well. This reduces the likelihood that you place a transaction just as the market is starting to get busy again that ends up being stuck for a very long time.

This PR also allows the specification of transaction confirmation target and whether the estimate should be conservative or not on a per-RPC call basis for sendtoaddress, sendmany, bumpfee, and fundrawtransaction using optional named arguments.

Left as an exercise to the reader is adding this to GUI functions to send transactions and bump fee.

morcos committed:
remove default argument from GetMinimumFee

morcos committed:
Introduce a fee estimate mode.

GetMinimumFee now passes the conservative argument into estimateSmartFee.
Call CalculateEstimateType(mode) before calling GetMinimumFee or estimateSmartFee to determine the value of this argument.
CCoinControl can now be used to control this mode.

morcos committed:
remove default argument from estimateSmartFee

morcos committed:
Change default fee estimation mode.

Fee estimates will default to be non-conservative if the transaction in question is opt-in-RBF.

morcos committed:
Add RPC options for RBF, confirmation target, and conservative fee estimation.

Add support for setting each of these attributes on a per RPC call basis to sendtoaddress, sendmany, fundrawtransaction (already had RBF), and bumpfee (already had RBF and conf target).

RHavar commented:
Concept ACK, this is really great and something that's been needed for a long time. It'll simplify some of the locking and `settxfee` stuff I've needed to do. 

But it'd be nice if it used an `options` object or something.  Otherwise if I want to specify the `conservative_estimate` I'll need to fill in all the other fields, which I might not want to know?

But I guess that's something that can be done in a different commit, with `if (arguments.length == 3 && typeof arguments == 'object')` style

instagibbs commented:
@RHavar I had the same exact reaction fwiw. `fee_options` or something.

morcos commented:
I thought the whole point of named arguments is we didn't have to worry about that any more.  I think I made sure these work correctly with holes so we can use them with named arguments.

RHavar commented:
Oh yeah, you're right. Guess I should learn how to use named arguments.

sipa commented:
Needs rebase.

morcos commented:
Rebased due to conflict with #10422

laanwj commented on src/wallet/rpcwallet.cpp:
```diff
@@ -909,7 +927,10 @@ UniValue sendmany(const JSONRPCRequest& request)
-            "\nResult:\n"
+            "6. opt_in_rbf             (boolean, optional) Allow this transaction to be replaced by a transaction with higher fees\n"
+            "7. conf_target            (numeric, optional) Confirmation target (in blocks)\n"
+            "8. conservative_estimate  (boolean, optional) Use conservative (potentially higher) fee estimation\n"
```
I'd prefer to use a string here (which maps to FeeEstimateMode) instead of a boolean. This would also allow adding additional fee estimation strategies in the future without changing the interface.

laanwj commented:
Concept ACK

morcos commented:
Switched to an estimate_mode string and squashed
5cbc1f6 ([rpcestimatechoice.ver2](https://github.com/morcos/bitcoin/commits/rpcestimatechoice.ver2)) -> 1bc6a4a ([rpcestimatechoices.ver2.squash](https://github.com/morcos/bitcoin/commits/rpcestimatechoice.ver2.squash))

morcos commented:
Changed argument name from opt_in_rbf to replaceable for conformance with bumpfee

instagibbs reviewed (commented):
utACK https://github.com/bitcoin/bitcoin/pull/10589/commits/ab07a1063ba495faa6029ebfc270e6c52d5a2cc9

morcos commented on src/wallet/wallet.cpp:
```diff
@@ -4145,3 +4148,14 @@ bool CMerkleTx::AcceptToMemoryPool(const CAmount& nAbsurdFee, CValidationState&
     return ::AcceptToMemoryPool(mempool, state, tx, true, NULL, NULL, false, nAbsurdFee);
 }
+
+bool CalculateEstimateType(FeeEstimateMode mode) {
```
https://github.com/bitcoin/bitcoin/pull/10586#discussion_r124370127

instagibbs replied:
If you change the semantics of the return value all the other callers will have to change too. Why not just have the caller consume the returned mode? Anyways, not a blocker.

instagibbs replied:
Ok we agree it should be a mode being returned ideally.

morcos replied:
Please see #10706 
CalculateEstimateType is now only used one place, in GetMinimumFee.

TheBlueMatt reviewed (commented):
I still dont really feel comfortable ACKing the Qt changes until I get a chance to test, but this is generally a very welcome improvement.

instagibbs commented:
needs rebase, will review

morcos commented:
rebased for adjacent line change in #10698 and removed a couple extra \n's 

no other changes

I believe we've settled on removing the CalculateEstimateType function entirely at the end of #10706 since it just becomes a helper for GetMinimumFee at that point.

TheBlueMatt reviewed (commented):
Looks good. Definitely want #10706 to go in with this.

morcos commented on src/qt/sendcoinsdialog.cpp:
```diff
@@ -166,6 +166,8 @@ void SendCoinsDialog::setModel(WalletModel *_model)
         connect(ui->checkBoxMinimumFee, SIGNAL(stateChanged(int)), this, SLOT(updateGlobalFeeVariables()));
         connect(ui->checkBoxMinimumFee, SIGNAL(stateChanged(int)), this, SLOT(coinControlUpdateLabels()));
+        connect(ui->optInRBF, SIGNAL(stateChanged(int)), this, SLOT(updateSmartFeeLabel()));
+        connect(ui->optInRBF, SIGNAL(stateChanged(int)), this, SLOT(coinControlUpdateLabels()));
```
Not sure if I understand your question (or how QT works) but I'm pretty sure it's correct as is.  Calling `setCheckState` below will trigger the signal if the checkbox changed from its initial value.

promag replied:
Either way, what about make an explicit call to `coinControlUpdateLabels()`?

morcos replied:
I don't think this is necessary.  It's analogous to setting the default sliderSmartFee done in the same section of code.

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -416,6 +416,12 @@ UniValue sendtoaddress(const JSONRPCRequest& request)
             "                             The recipient will receive less bitcoins than you enter in the amount field.\n"
+            "6. replaceable            (boolean, optional) Allow this transaction to be replaced by a transaction with higher fees via BIP 125\n"
+            "7. conf_target            (numeric, optional) Confirmation target (in blocks)\n"
+            "8. \"estimate_mode\"      (string, optional, default=UNSET) The fee estimate mode, must be one of:\n"
```
`fee_estimate_mode`, `fee_mode` or `estimate_mode`?

morcos replied:
Hmm..  Yes, I think it might make sense to change this to `fee_mode`, but leave it `estimate_mode` for estimateSmartFee (done in another PR).  Even though they take the same values now, I could imagine they would differ in the future.  Any other thoughts?

promag replied:
I just wanted to point out the inconsistencies between arguments, variable names and types.

morcos replied:
I decided leaving it as estimate_mode in both RPC's is preferable. 
The inconsistency between variable and parameter name is not worth the churn imo.

promag commented on src/wallet/coincontrol.h:
```diff
@@ -26,6 +27,8 @@ class CCoinControl
     //! Signal BIP-125 replace by fee.
     bool signalRbf;
+    //! Fee estimation mode to control arguments to estimateSmartFee
+    FeeEstimateMode m_fee_mode;
```
`m_fee_estimate_mode`?

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -401,7 +401,7 @@ UniValue sendtoaddress(const JSONRPCRequest& request)
-    if (request.fHelp || request.params.size() < 2 || request.params.size() > 5)
+    if (request.fHelp || request.params.size() < 2 || request.params.size() > 8)
         throw std::runtime_error(
             "sendtoaddress \"address\" amount ( \"comment\" \"comment_to\" subtractfeefromamount )\n"
```
Missing new arguments.

morcos replied:
will do

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -910,7 +935,13 @@ UniValue sendmany(const JSONRPCRequest& request)
             "      ,...\n"
             "    ]\n"
-            "\nResult:\n"
+            "6. replaceable            (boolean, optional) Allow this transaction to be replaced by a transaction with higher fees via BIP 125\n"
```
Add `options` instead of 3 more optional arguments?

promag replied:
I saw the above comment about named arguments. IMO from the client side feels better as an option.

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -416,6 +416,12 @@ UniValue sendtoaddress(const JSONRPCRequest& request)
             "                             transaction, just kept in your wallet.\n"
             "5. subtractfeefromamount  (boolean, optional, default=false) The fee will be deducted from the amount being sent.\n"
             "                             The recipient will receive less bitcoins than you enter in the amount field.\n"
+            "6. replaceable            (boolean, optional) Allow this transaction to be replaced by a transaction with higher fees via BIP 125\n"
```
Add `options` instead of 3 more optional arguments?

morcos replied:
I believe named arguments are preferable

promag replied:
I understand, but for those that use indexed arguments, they have to pass the middle arguments.

promag commented on src/wallet/wallet.cpp:
```diff
@@ -4154,3 +4157,15 @@ bool CMerkleTx::AcceptToMemoryPool(const CAmount& nAbsurdFee, CValidationState&
     return ::AcceptToMemoryPool(mempool, state, tx, true, NULL, NULL, false, nAbsurdFee);
 }
+
+bool CalculateEstimateType(FeeEstimateMode mode, bool opt_in_rbf) {
```
`IsFeeEstimateConservative`?

morcos replied:
Function removed in #10706 (or will be shortly)

promag replied:
Still, name not clear don't you think?

morcos replied:
yeah but who cares, its going away completely in a couple of commits

promag reviewed (commented):
IMO this would be a cleaner implementation
```cpp
bool FeeEstimateModeFromString(const std::string& string, FeeEstimateMode& fee_estimate_mode)
{
    // lookup string
    return false;
}

...


if (!FeeEstimateModeFromString(request.params[7].get_str(), coin_control.m_fee_mode)) {
    throw JSONRPCError(RPC_INVALID_PARAMETER, ...);
}
```

morcos commented:
Took some of the nits
([rpcestimatechoice.ver3](https://github.com/morcos/bitcoin/commits/rpcestimatechoice.ver3)) -> f135923 ([rpcestimatechoices.ver3.squash](https://github.com/morcos/bitcoin/commits/rpcestimatechoice.ver3.squash))

ryanofsky reviewed (commented):
utACK f135923ee2cf1a1a9a436626dc5b9219f8ad97da

laanwj commented:
utACK f135923
